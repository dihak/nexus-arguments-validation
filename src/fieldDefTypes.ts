import { printedGenTyping, printedGenTypingImport } from 'nexus/dist/core'
import { Schema } from 'yup'

const FieldValidationResolverImport = printedGenTypingImport({
  module: 'nexus-arguments-validation/dist/fieldDefTypes',
  bindings: ['FieldValidationResolver'],
})

export const fieldDefTypes = printedGenTyping({
  optional: true,
  name: 'validationSchema', 
  description:
    'Yup schema or function that returns a Yup schema. This is used for input validation.',
  type: 'FieldValidationResolver',
  imports: [FieldValidationResolverImport],
})

export type FieldValidationResolver = Schema<any> | (() => Schema<any>)