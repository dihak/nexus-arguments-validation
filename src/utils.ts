import setIn from 'set-value'
import { ValidationError } from 'yup'

export type Maybe<T> = T | null | undefined

export const formatValidationError = (error: ValidationError) => {
  return error.inner.reduce((acc, val) => {
    setIn(acc, val.path, val.message)
    return acc
  }, {})
}
