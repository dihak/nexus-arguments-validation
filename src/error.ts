import { GraphQLError, Source, SourceLocation, ASTNode } from 'graphql'
import { Maybe, formatValidationError } from './utils'
import { ValidationError } from 'yup'

export class ArgumentsValidationError extends Error implements GraphQLError {
  public extensions: Record<string, any>
  public message: string
  readonly locations: ReadonlyArray<SourceLocation> | undefined
  readonly path: ReadonlyArray<string | number> | undefined
  readonly nodes: ReadonlyArray<ASTNode> | undefined
  readonly source: Source | undefined
  readonly positions: ReadonlyArray<number> | undefined
  readonly originalError: Maybe<Error>

  constructor(
    message: string,
    error: ValidationError,
    code: string = 'ARGUMENTS_VALIDATION_ERROR',
  ) {
    super()

    this.message = message
    this.extensions = {
      code,
      argumentsValidations: formatValidationError(error),
    }
  }
}
