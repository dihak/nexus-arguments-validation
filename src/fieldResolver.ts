import type { ObjectSchema, ValidationError } from 'yup'
import { MiddlewareFn, plugin, CreateFieldResolverInfo } from '@nexus/schema/dist/core'
import { ArgumentsValidationError } from './error'
import { FieldValidationResolver } from './fieldDefTypes'

export const fieldResolver = (
  config: CreateFieldResolverInfo,
): MiddlewareFn | undefined => {
  const validationConfig = config.fieldConfig.extensions?.nexus?.config

  // If the field doesn't have validation schema field, skip wrapping resolver
  if (validationConfig.validationSchema === undefined) {
    return
  }

  let {
    validationSchema,
  }: { validationSchema: FieldValidationResolver } = validationConfig

  if (typeof validationSchema === 'function') {
    validationSchema = validationSchema()
  }

  // Check if its Yup schema with `validate` function
  if (typeof validationSchema !== 'object' || !validationSchema.validate) {
    // console.log(validationSchema, config.fieldConfig)
    console.error(
      new Error(
        `The validationSchema property provided to ${
          validationConfig.type
        } should be a Yup schema, saw ${typeof validationSchema}`,
      ),
    )
    return
  }

  return (root, args, ctx, info, next): MiddlewareFn => {
    let validationResult
    try {
      validationResult = (validationSchema as ObjectSchema<any>).validate(
        args,
        { abortEarly: false },
      )
    } catch (e) {
      validationResult = Promise.reject(e)
    }

    return plugin.completeValue(
      validationResult,
      () => {
        return next(root, args, ctx, info)
      },
      (err: ValidationError) => {
        throw new ArgumentsValidationError('Bad user input', err)
      },
    )
  }
}
