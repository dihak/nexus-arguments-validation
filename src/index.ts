import { plugin } from 'nexus'
import { fieldDefTypes } from './fieldDefTypes'
import { fieldResolver } from './fieldResolver'

export const nexusArgumentsValidationPlugin = () => {
  return plugin({
    name: 'nexus-arguments-validation',
    description: 'Validation of arguments with Yup object shape.',
    fieldDefTypes,
    onCreateFieldResolver: config => fieldResolver(config),
  })
}
