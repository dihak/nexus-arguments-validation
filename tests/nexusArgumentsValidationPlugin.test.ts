import {
  makeSchema,
  objectType,
  queryField,
  mutationField,
  stringArg,
} from '@nexus/schema'
import { nexusArgumentsValidationPlugin } from '../src'
import { graphql } from 'graphql'
import { object, string } from 'yup'

describe('nexusInputValidationPlugin', () => {
  const consoleError = jest.spyOn(console, 'error').mockImplementation(() => {})

  afterEach(() => {
    jest.resetAllMocks()
  })

  const UserValidation = object().shape({
    email: string().email('Email is not in valid format!'),
    password: string().min(8, 'Password too short!'),
  })

  const types = [
    objectType({
      name: 'User',
      definition: t => {
        t.string('email')
        t.string('password')
      },
    }),
    queryField('user', {
      type: 'User',
      resolve: () => {
        return { email: 'input@plugin.com', password: 'password' }
      },
    }),
    mutationField('signup', {
      type: 'User',
      args: {
        email: stringArg(),
        password: stringArg(),
      },
      // @ts-ignore
      validationSchema: UserValidation,
      resolve: (_, { email, password }) => {
        return { email, password }
      },
    }),
    mutationField('callback', {
      type: 'User',
      args: {
        email: stringArg(),
        password: stringArg(),
      },
      // @ts-ignore
      validationSchema: () => UserValidation,
      resolve: (_, { email, password }) => {
        return { email, password }
      },
    }),
  ]

  const nexusSchema = makeSchema({
    types,
    outputs: false,
    plugins: [nexusArgumentsValidationPlugin()],
  })

  const query = (field: string) => {
    return graphql(
      nexusSchema,
      `{
        ${field}{
          email
        }
      }`,
    )
  }

  const mutation = (
    field: string,
    variables: { [key: string]: any },
    schema = nexusSchema,
  ) => {
    return graphql(
      schema,
      `mutation ($email: String! $password: String!){
        ${field}(email: $email, password: $password){ 
          email
          password
        }
      }`,
      {},
      {},
      variables,
    )
  }

  test('query user', async () => {
    const { data, errors = [] } = await query('user')
    expect(consoleError).not.toHaveBeenCalled()
    expect(data!.user.email).toBe('input@plugin.com')
    expect(errors.length).toBe(0)
  })

  test('mutation with validationSchema object', async () => {
    const variables = {
      email: 'input@plugin.com',
      password: 'password',
    }
    const { data, errors = [] } = await mutation('signup', variables)

    expect(consoleError).not.toHaveBeenCalled()
    expect(data!.signup).toEqual({
      email: 'input@plugin.com',
      password: 'password',
    })
    expect(errors.length).toBe(0)
  })

  test('mutation with validationSchema object callback', async () => {
    const variables = {
      email: 'input@plugin.com',
      password: 'password',
    }
    const { data, errors = [] } = await mutation('callback', variables)

    expect(consoleError).not.toHaveBeenCalled()
    expect(data!.callback).toEqual({
      email: 'input@plugin.com',
      password: 'password',
    })
    expect(errors.length).toBe(0)
  })

  test('error validation with single input field', async () => {
    const variables = {
      email: 'input@plugin.com',
      password: 'pass',
    }
    const { data, errors = [] } = await mutation('signup', variables)

    expect(consoleError).not.toHaveBeenCalled()
    expect(data).toBe(null)
    expect(errors.length).toBe(1)
    expect(errors[0].message).toBe('Bad user input')
    expect(errors[0].extensions!.code).toBe('ARGUMENTS_VALIDATION_ERROR')
    expect(errors[0].extensions!.argumentsValidations).toEqual({
      password: 'Password too short!',
    })
  })

  test('error validation with multiple input field', async () => {
    const variables = {
      email: 'input@plugin',
      password: 'pass',
    }
    const { data, errors = [] } = await mutation('signup', variables)

    expect(consoleError).not.toHaveBeenCalled()
    expect(data).toBe(null)
    expect(errors.length).toBe(1)
    expect(errors[0].message).toBe('Bad user input')
    expect(errors[0].extensions!.code).toBe('ARGUMENTS_VALIDATION_ERROR')
    expect(errors[0].extensions!.argumentsValidations).toEqual({
      email: 'Email is not in valid format!',
      password: 'Password too short!',
    })
  })

  test('incorrect validationSchema shape', async () => {
    const variables = {
      email: 'input@plugin.com',
      password: 'password',
    }

    const nexusSchema = makeSchema({
      types: [
        ...types,
        mutationField('wrongValidationSchema', {
          type: 'User',
          args: {
            email: stringArg({ required: true }),
            password: stringArg(),
          },
          // @ts-ignore
          validationSchema: () => ({}),
          resolve: async (_, { email, password }) => {
            return { email, password }
          },
        }),
      ],
      outputs: false,
      plugins: [nexusArgumentsValidationPlugin()],
    })

    const { data, errors = [] } = await mutation(
      'wrongValidationSchema',
      variables,
      nexusSchema,
    )

    expect(consoleError.mock.calls[0]).toMatchSnapshot()
    expect(errors.length).toEqual(0)
    expect(data!.wrongValidationSchema).toEqual(variables)
  })

  // test('signup user without errors', async () => {
  //   const variables = {
  //     email: 'input@plugin.com',
  //     password: 'password',
  //   }
  //   const { data, errors = [] } = await mutation(
  //     'wrongValidationSchema',
  //     variables
  //   )

  //   expect(consoleError).not.toHaveBeenCalled();
  //   console.log(data, errors)
  // })

  // Console warn
})
