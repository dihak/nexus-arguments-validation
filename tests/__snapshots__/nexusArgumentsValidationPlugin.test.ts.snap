// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nexusInputValidationPlugin incorrect validationSchema shape 1`] = `
Array [
  [Error: The validationSchema property provided to User should be a Yup schema, saw object],
]
`;
